test/integration/cds_integration_test.cc

```diff
@@ -222,6 +222,117 @@ TEST_P(CdsIntegrationTest, CdsClusterTeardownWhileConnecting) {
   EXPECT_LE(cx_counter->value(), 1);
 }
 
+class LazyInitClusterStatsTest : public CdsIntegrationTest {
+
+public:
+  void initialize() {
+    use_real_stats_ = true;
+    config_helper_.addConfigModifier([](::envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      bootstrap.set_enable_lazyinit_stats(true);
+    });
+    CdsIntegrationTest::initialize();
+    test_server_->waitForCounterGe("cluster_manager.cluster_added", 1);
+  }
+
+  void sendRequestToCluster1AndWaitForResponse() {
+    BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+        lookupPort("http"), "GET", "/cluster1", "", downstream_protocol_, version_, "foo.com");
+    ASSERT_TRUE(response->complete());
+    cleanupUpstreamAndDownstream();
+  };
+
+  void updateCluster1() {
+    envoy::config::cluster::v3::Cluster cluster1_updated = cluster_creator_(
+        ClusterName1, fake_upstreams_[UpstreamIndex2]->localAddress()->ip()->port(),
+        Network::Test::getLoopbackAddressString(ipVersion()), "ROUND_ROBIN");
+
+    sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+        Config::TypeUrl::get().Cluster, {cluster1_updated}, {cluster1_updated}, {}, "42");
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsClientTypeDelta, LazyInitClusterStatsTest,
+                         DELTA_SOTW_GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// Test that LazyInitTrafficStats gets created and updated correctly.
+TEST_P(LazyInitClusterStatsTest, LazyInitTrafficStatsWithClusterCreateUpdateDelete) {
+  initialize();
+
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total"), nullptr);
+
+  sendRequestToCluster1AndWaitForResponse();
+  // Cluster_1 trafficStats updated.
+  EXPECT_EQ(test_server_->gauge("cluster.cluster_1.ClusterTrafficStats.initialized")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+
+  updateCluster1();
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 2);
+
+  // Now the ClusterTrafficStats 'initialized' gauge >= 1, since there is a small race window
+  // between the testing thread fetching the value before the old version ClusterTrafficStats gets
+  // deleted from main thread and worker thread.
```

Comment:
I commented on this earlier but I think it got lost in github.

We should ensure we run the race so it wins each way, using thread-synchronizer, if necessary.

Reply:
I think we talked about this before, and we agreed that too much
The race window is that: 1. in this test, we are making a read on the cluster_1.ClusterTrafficStats.initialized gauge,
2. in either some worker thread, or the main thread, the threadlocal cluster destructor eventually removed the last reference to the shared_ptr clusterInfoImpl.

I am not sure if we should latch a checkpoint into lazyinit to capture this, since it's gonna be another unique_ptr cost against the per cluster/route/vhost StatsStruct.

thoughts?

Reply:
Yeah let's tackle reproducing this race in the not-yet-existing lazy_init_test.cc. I should've insisted on a unit test for the new class earlier; it's going to be much easier to robustly test it on its own.